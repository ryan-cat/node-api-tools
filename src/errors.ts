import { createError, ApolloError } from 'apollo-errors';
import { ErrorInfo, formatError as formatApolloError, isInstance as isApolloErrorInstance } from 'apollo-errors';
import { GraphQLError } from 'graphql';

export const AuthenticationError = createError('AuthenticationError', {
  message: 'You must be authenticated to access this data.'
});

class BaseError extends ApolloError {
  public data: { [key: string]: string[] };
}

export class ItemNotFoundError extends BaseError {
  private static DEFAULT_MESSAGE = 'The item could not be found.';

  constructor(item: string, action: string = null) {
    const defaultConfig = { message: ItemNotFoundError.DEFAULT_MESSAGE };

    super('UserInputError', defaultConfig, defaultConfig);

    const ending = action ? ` to ${action}.` : '.';
    this.message = `The ${item} could not be found${ending}`;
  }
}

export class UserInputError extends BaseError {
  private static DEFAULT_MESSAGE = 'The provided input is invalid.';

  constructor(message: string) {
    const defaultConfig = { message: UserInputError.DEFAULT_MESSAGE };

    super('UserInputError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export class ValidationError extends BaseError {
  private static DEFAULT_MESSAGE = 'The item could not be saved due to issues with the provided input.';

  constructor(item: string, action = 'saved', data: any = {}) {
    const defaultConfig = { message: ValidationError.DEFAULT_MESSAGE };

    super('ValidationError', defaultConfig, defaultConfig);

    this.message = `${item} could not be ${action} due to issues with the provided input.`;
    this.data = data;
  }
}

export class AuthorizationError extends BaseError {
  private static DEFAULT_MESSAGE = 'You do not have permission to perform this task.';

  constructor(message: string) {
    const defaultConfig = { message: AuthorizationError.DEFAULT_MESSAGE };

    super('AuthorizationError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export class AuthorizationActionError extends BaseError {
  private static DEFAULT_MESSAGE = 'You do not have permission to perform this task.';

  constructor(action: string) {
    const defaultConfig = { message: AuthorizationActionError.DEFAULT_MESSAGE };

    super('AuthorizationError', defaultConfig, defaultConfig);

    this.message = `You do not have permission to ${action}.`;
  }
}

export class InternalServerError extends BaseError {
  private static DEFAULT_MESSAGE = 'Oops! Something went wrong. Please try again.';

  constructor(message: string = null) {
    const defaultConfig = { message: InternalServerError.DEFAULT_MESSAGE };
    super('InternalServerError', defaultConfig, defaultConfig);

    this.message = message || this.message;
  }
}

export class APIError extends BaseError {
  private static DEFAULT_MESSAGE = 'The parameter configuration provided is invalid.';

  constructor(message: string = null) {
    const defaultConfig = { message: APIError.DEFAULT_MESSAGE };
    super('APIError', defaultConfig, defaultConfig);

    this.message = message || this.message;
  }
}

export const formatError = (error: GraphQLError, mask: boolean): any => {
  const err = error.originalError;

  if (isApolloErrorInstance(err)) {
    return formatApolloError(err);
  } else if (err && !isApolloErrorInstance(error) && !mask) {
    return formatApolloError(new InternalServerError(err.message));
  } else {
    return formatApolloError(new InternalServerError());
  }
};
