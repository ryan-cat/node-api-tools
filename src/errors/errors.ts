import { ApolloError, formatError as formatApolloError, isInstance as isApolloErrorInstance } from 'apollo-errors';
import { GraphQLError, GraphQLFormattedError } from 'graphql';

class BaseError extends ApolloError {
  public data: any;
}

export class AuthenticationError extends BaseError {
  private static DEFAULT_MESSAGE = 'You must be authenticated to access this data.';

  constructor() {
    const defaultConfig = { message: AuthenticationError.DEFAULT_MESSAGE };

    super('AuthenticationError', defaultConfig, defaultConfig);
  }
}

export class ItemNotFoundError extends BaseError {
  private static DEFAULT_MESSAGE = 'The item could not be found.';

  constructor(item: string, action: string = null) {
    const defaultConfig = { message: ItemNotFoundError.DEFAULT_MESSAGE };

    super('UserInputError', defaultConfig, defaultConfig);

    const ending = action ? ` to ${action}.` : '.';
    this.message = `The ${item} could not be found${ending}`;
  }
}

export class UserInputError extends BaseError {
  private static DEFAULT_MESSAGE = 'The provided input is invalid.';

  constructor(message: string = UserInputError.DEFAULT_MESSAGE) {
    const defaultConfig = { message: UserInputError.DEFAULT_MESSAGE };

    super('UserInputError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export interface ValidationErrorItem {
  field: (string | number)[];
  error: string;
}

export class ValidationError extends BaseError {
  private static DEFAULT_MESSAGE = 'The item could not be saved due to issues with the provided input.';

  constructor(item: string, action = 'saved', data: ValidationErrorItem[] = null) {
    const defaultConfig = { message: ValidationError.DEFAULT_MESSAGE };

    super('ValidationError', defaultConfig, defaultConfig);

    this.message = `${item} could not be ${action} due to issues with the provided input.`;
    this.data = data;
  }
}

export class AuthorizationError extends BaseError {
  private static DEFAULT_MESSAGE = 'You do not have permission to perform this task.';

  constructor(message: string = AuthorizationError.DEFAULT_MESSAGE) {
    const defaultConfig = { message: AuthorizationError.DEFAULT_MESSAGE };

    super('AuthorizationError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export class AuthorizationActionError extends BaseError {
  private static DEFAULT_MESSAGE = 'You do not have permission to perform this task.';

  constructor(action: string = AuthorizationActionError.DEFAULT_MESSAGE) {
    const defaultConfig = { message: AuthorizationActionError.DEFAULT_MESSAGE };

    super('AuthorizationError', defaultConfig, defaultConfig);

    this.message = `You do not have permission to ${action}.`;
  }
}

export class InternalServerError extends BaseError {
  private static DEFAULT_MESSAGE = 'Oops! Something went wrong. Please try again.';

  constructor(message: string = InternalServerError.DEFAULT_MESSAGE) {
    const defaultConfig = { message: InternalServerError.DEFAULT_MESSAGE };
    super('InternalServerError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export class APIError extends BaseError {
  private static DEFAULT_MESSAGE = 'The parameter configuration provided is invalid.';

  constructor(message: string = APIError.DEFAULT_MESSAGE) {
    const defaultConfig = { message: APIError.DEFAULT_MESSAGE };
    super('APIError', defaultConfig, defaultConfig);

    this.message = message;
  }
}

export const formatError = (error: GraphQLError, mask: boolean = true): GraphQLFormattedError<Record<string, any>> => {
  const err = error?.originalError;

  let newError = formatApolloError(new InternalServerError());
  if (isApolloErrorInstance(err)) {
    newError = formatApolloError(err);
  } else if (err && !isApolloErrorInstance(error) && !mask) {
    newError = formatApolloError(new InternalServerError(err.message));
  }

  return {
    message: newError.message,
    locations: newError.locations,
    extensions: {
      code: newError.name,
      data: newError.data,
      timeThrown: newError.time_thrown
    }
  } as GraphQLFormattedError<Record<string, any>>;
};
