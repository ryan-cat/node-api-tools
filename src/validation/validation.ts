import { ValidationError } from '../errors';
import { Schema, ValidationErrorItem } from 'joi';

export interface ValidationMessageOptions {
  item: string;
  action?: string;
}

const defaultValidationOptions: ValidationMessageOptions = {
  item: 'Item',
  action: 'saved'
};

export const validate = <E>(
  item: E,
  validator: Schema,
  validationOptions: ValidationMessageOptions = defaultValidationOptions,
  throwError = true
): ValidationError => {
  if (validator == null) {
    throw new Error('Provided validator is null.');
  }

  const result = validator.validate(item, {
    messages: {
      key: '{{label}} '
    },
    convert: false,
    abortEarly: false,
    allowUnknown: true
  });

  if (result && result.error) {
    const data = {};

    result.error.details.forEach((detail) => {
      const key = detail.context.key;
      const message = detail.message.charAt(0).toUpperCase() + detail.message.slice(1) + '.';

      if (!data[key]) {
        data[key] = [message];
      } else {
        data[key] = [...data[key], message];
      }
    });

    const options = validationOptions || defaultValidationOptions;

    if (throwError) {
      throw new ValidationError(options.item, options.action, data);
    } else {
      return new ValidationError(options.item, options.action, data);
    }
  }
};

export abstract class Validatable {
  public validator: Schema;
  public validationOptions: ValidationMessageOptions;

  public validate = <E>(item: E, throwError = true) => {
    return validate(item, this.validator, this.validationOptions, throwError);
  };
}

export const mergeValidationErrors = (err1: ValidationError, err2: ValidationError, throwError = false): ValidationError => {
  let err = err1;
  if (!err1) {
    err = err2;
  } else if (err2) {
    Object.keys(err2.data).forEach((key) => {
      if (err1.data[key] && err1.data[key].length > 0) {
        err.data[key] = [...err2.data[key], ...err1.data[key]];
      } else {
        err.data[key] = err2.data[key];
      }
    });
  }

  if (err && throwError) {
    throw err;
  } else {
    return err;
  }
};

export const customValidationMessages = (errors: ValidationErrorItem[], map: { [key: string]: string }): ValidationErrorItem[] => {
  errors.forEach((err) => {
    if (map[err.type]) {
      err.message = map[err.type];
    }
  });

  return errors;
};
