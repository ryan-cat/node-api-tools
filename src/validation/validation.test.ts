import { ValidationError } from './../errors/errors';
import { validate, mergeValidationErrors, Validatable, customValidationMessages, ValidationMessageOptions } from './index';
import * as validation from './validation';
import * as joi from 'joi';

describe('validate', () => {
  const options: ValidationMessageOptions = {
    item: 'User',
    action: 'created'
  };

  const schema: joi.Schema = joi.object();

  const validationResult = {
    error: {
      details: [
        {
          context: {
            key: 'email'
          },
          message: 'Invalid email'
        },
        {
          context: {
            key: 'email'
          },
          message: 'Email too long'
        },
        {
          context: {
            key: 'username'
          },
          message: 'Username too long'
        }
      ]
    }
  };

  const expectedErrorData = {
    email: [validationResult.error.details[0].message + '.', validationResult.error.details[1].message + '.'],
    username: [validationResult.error.details[2].message + '.']
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should throw error if validator is null', () => {
    try {
      validate({}, null);
    } catch (err) {
      return;
    }

    fail(new Error('Expected error to be throw because validator is null.'));
  });

  it('should validate and return null for no issues', () => {
    jest.spyOn(schema, 'validate').mockImplementationOnce(() => {});

    const result = validate({}, schema);
    expect(result).toBeUndefined();
  });

  it('should validate and throw error due to validation issues', () => {
    jest.spyOn(schema, 'validate').mockImplementationOnce(() => validationResult);

    try {
      validate({}, schema);
    } catch (err) {
      const error = err as ValidationError;
      expect(error.message).toMatchSnapshot();
      expect(error.data).toStrictEqual(expectedErrorData);

      return;
    }

    fail(new Error('Expected validation error to be thrown.'));
  });

  it('should validate and throw error due to validation issues with custom message options', () => {
    jest.spyOn(schema, 'validate').mockImplementationOnce(() => validationResult);

    try {
      validate({}, schema, options);
    } catch (err) {
      const error = err as ValidationError;
      expect(error.message).toMatchSnapshot();
      expect(error.data).toStrictEqual(expectedErrorData);

      return;
    }

    fail(new Error('Expected validation error to be thrown.'));
  });

  it('should validate and return error due to validation issues', () => {
    jest.spyOn(schema, 'validate').mockImplementationOnce(() => validationResult);

    const error = validate({}, schema, null, false);
    expect(error.message).toMatchSnapshot();
    expect(error.data).toStrictEqual(expectedErrorData);
  });

  it('should call validate for validatable class', () => {
    class User extends Validatable {
      validator = schema;
    }

    const spy = jest.spyOn(validation, 'validate');

    const u = new User();
    u.validate({});

    expect(spy).toBeCalledTimes(1);
    expect(spy).toBeCalledWith({}, schema, undefined, true);
  });
});

describe('merge validation errors', () => {
  let error1;
  let error2;
  let mergedErrorData;

  beforeEach(() => {
    error1 = new ValidationError('User', null, {
      email: ['test', 'sample'],
      username: ['max', 'format']
    });

    error2 = new ValidationError('User', null, {
      email: ['go', 'sample'],
      name: ['try']
    });

    mergedErrorData = {
      email: ['go', 'sample', 'test', 'sample'],
      username: ['max', 'format'],
      name: ['try']
    };
  });

  it('should return first error when second error is null', () => {
    const result = mergeValidationErrors(error1, null);
    expect(result).toStrictEqual(error1);
  });

  it('should return second error when first error is null', () => {
    const result = mergeValidationErrors(null, error1);
    expect(result).toStrictEqual(error1);
  });

  it('should merge and throw validation errors', () => {
    try {
      mergeValidationErrors(error1, error2, true);
    } catch (err) {
      const error = err as ValidationError;
      expect(error.message).toBe(error1.message);
      expect(error.data).toStrictEqual(mergedErrorData);

      return;
    }

    fail(new Error('Expected validation error to be thrown.'));
  });

  it('should merge and return validation errors', () => {
    const result = mergeValidationErrors(error1, error2);
    expect(result.message).toBe(error1.message);
    expect(result.data).toStrictEqual(mergedErrorData);
  });
});

describe('custom validation errors', () => {
  it('should replace default error message with custom error message', () => {
    const errors: joi.ValidationErrorItem[] = [
      { type: 'maxlength', message: 'The length of the item is too long.', path: [] },
      { type: 'minlength', message: 'The length of the item is too short.', path: [] }
    ];

    const map = {
      minlength: "The item's lenghth is not long enough."
    };

    const result = customValidationMessages(errors, map);

    expect(result).toHaveLength(2);
    expect(result[0].message).toBe(errors[0].message);
    expect(result[1].message).toBe(map.minlength);
  });
});
