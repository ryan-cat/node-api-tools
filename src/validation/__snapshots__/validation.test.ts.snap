// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate should validate and return error due to validation issues 1`] = `"Item could not be saved due to issues with the provided input."`;

exports[`validate should validate and return error due to validation issues 2`] = `
Object {
  "email": Array [
    "\\"email\\" must be a valid email.",
  ],
  "username": Array [
    "\\"username\\" length must be less than or equal to 10 characters long.",
  ],
}
`;

exports[`validate should validate and throw error due to validation issues 1`] = `"Item could not be saved due to issues with the provided input."`;

exports[`validate should validate and throw error due to validation issues 2`] = `
Object {
  "email": Array [
    "Invalid email.",
  ],
  "username": Array [
    "Username too long.",
  ],
}
`;

exports[`validate should validate and throw error due to validation issues with custom message options 1`] = `"User could not be created due to issues with the provided input."`;

exports[`validate should validate and throw error due to validation issues with custom message options 2`] = `
Object {
  "email": Array [
    "\\"email\\" must be a valid email.",
  ],
  "username": Array [
    "\\"username\\" length must be less than or equal to 10 characters long.",
  ],
}
`;
